src/processor/document/candidate_relation_adder.py

```diff
@@ -0,0 +1,183 @@
+from __future__ import annotations
+
+import json
+import logging
+import random
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import Any, Dict, List, overload
+
+from pytorch_ie.annotations import BinaryRelation, LabeledSpan, Span
+from pytorch_ie.core import AnnotationList, annotation_field
+from pytorch_ie.documents import TextDocument
+from pytorch_ie.utils.span import is_contained_in
+
+from pie_utils.span.slice import distance
+from pie_utils.statistics import WithStatistics
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class DocumentWithEntitiesRelationsAndPartition(TextDocument):
+    entities: AnnotationList[LabeledSpan] = annotation_field(target="text")
+    relations: AnnotationList[BinaryRelation] = annotation_field(target="entities")
+    partition: AnnotationList[Span] = annotation_field(target="text")
+
+
+class CandidateRelationAdder(WithStatistics):
+    def reset_statistics(self):
+        self._statistics: dict[str, Any] = {
+            "num_total_relation_candidates": 0,
+            "num_available_relations": 0,
+            "available_rels_within_allowed_distance": 0,
+            "num_added_relation_not_taken": 0,
+            "num_rels_within_allowed_distance": 0,
+            "num_candidates_not_taken": defaultdict(int),
+            "distances_taken": defaultdict(list),
+        }
+
+    def show_statistics(self, description: str | None = None):
+        description = description or "Statistics"
+        logger.info(f"{description}: \n{json.dumps(self._statistics, indent=2)}")
+
+    @overload
+    def update_statistics(self, key: str, value: int):
+        pass
+
+    @overload
+    def update_statistics(self, key: str, value: dict[str, list]):
+        pass
+
+    @overload
+    def update_statistics(self, key: str, value: dict[str, int]):
+        pass
+
+    def update_statistics(self, key: str, value: int | dict[str, list] | dict[str, int]):
+        if isinstance(value, int):
+            self._statistics[key] += value
+        elif isinstance(value, Dict):
+            for k, v in value.items():
+                if isinstance(v, List):
+                    self._statistics[key][k] += v
+                elif isinstance(v, int):
+                    self._statistics[key][k] += v
+                else:
+                    raise TypeError(
+                        f"type of given key [{type(key)}] or value [{type(value)}] is incorrect."
+                    )
+        else:
+            raise TypeError(
+                f"type of given key [{type(key)}] or value [{type(value)}] is incorrect."
+            )
```

Comment:
What's this all about? I thought this is handled by `@overload`?

Reply:
Overloading using tying is done in this manner that we define possible variation of method but do not implement in those method and then we create a final method where we implement everything. See here: https://adamj.eu/tech/2021/05/29/python-type-hints-how-to-use-overload/

Reply:
thanks for the education :)

Reply:
However, is overloading required at all then?

Reply:
I guess it is in a way overloading and @overloads is just to signify that there are different interpretation of the method. Though I checked without having those defined method, it works. Shall I maybe remove those definitions then ?

Reply:
yeah, because as stated in the introduction of the article you have linked, overloading seems to be just necessary to restrict for certain combinations of input + output types. Since out method does not return anything, I think we do not need it. (I had another (wrong) understanding of `@overload` in mind when proposing to use it)